Обратная сторона реальности

Ivan-Pelipenko opened the pull request:
- Данные предложений
- Данные отзывов

---
:mortar_board: [Обратная сторона реальности](https://up.htmlacademy.ru/react/12/user/2176283/tasks/7)

Ivan-Pelipenko committed:
Обратная сторона реальности

- Данные предложений
- Данные отзывов

Ivan-Pelipenko committed:
- Подключим 'offers' в  index.tsx
- Передадим 'offers' через пропсы в компонент `App`

Ivan-Pelipenko committed:
- передаем данные в компонент "Карточка предложения"

Ivan-Pelipenko committed:
- передаем в компонент предложения данные.

Ivan-Pelipenko committed:
-Отрисовка моковых предложений на главной странице с помощью компонента "Список предложений"

Ivan-Pelipenko committed:
- Активная карточка предложения (наведение курсора)

saimon395 commented on project/src/pages/welcome-page/welcome-page.tsx:
```diff
@@ -113,11 +116,9 @@ function WelcomePage({offerCount}: WelcomePageProps): JSX.Element {
-              </div>
+              {
+                <OffersList offers={offers} offerCount={offerCount}/>
+              }
```
когда компонент подключаешь фигурные скобки не нужны 
```suggestion
              
                <OffersList offers={offers} offerCount={offerCount}/>
              
```

saimon395 commented on project/src/pages/welcome-page/welcome-page.tsx:
```diff
@@ -1,11 +1,14 @@
 import { Helmet } from 'react-helmet-async';
-import OfferCard from '../../components/card';
+//import OfferCard from '../../components/card';
+import OffersList from '../offers-list/offers-list';
```
закоментированный код лучше сразу удаляй

saimon395 commented on project/src/pages/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,22 @@
+
+function OffersList(props: OffersListProps): JSX.Element {
+  const {offers} = props;
+  const {offerCount} = props;
```
лучше сразу деструктурируй 
```suggestion
function OffersList({ offerCount, offers }: OffersListProps): JSX.Element {
```

saimon395 commented on project/src/pages/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,22 @@
+      {
+        template.slice(0,offerCount)
+      }
+    </div>);
```
здесь тоже можно отдельно переменную не создавать а писать map прямо в jsx
```suggestion
  return (
    <div className="cities__places-list places__list tabs__content">
      { 
          offers.slice(0,offerCount).map((offer) => <OfferCard key={offer.id} offer={offer}/>)
      }
    </div>);
```

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -1,7 +1,100 @@
+  };
+
+  const collage = () => {
+    if (images.length) {
+      const template = images.map((image, id) => (
+        <div key={`${id}-${image}`} className="property__image-wrapper">
+          <img className="property__image" src={image} alt="studio"/>
+        </div>
+      ));
+      return template.slice(0, 6);
+    }
+  };
```
не надо так делать, лучше сразу в отдельные компоненты выноси, что бы этот файл не захламлять

saimon395 replied:
с optionsList, statusPro, premiumSticker  и reviewsList аналогично

saimon395 commented on project/src/components/card.tsx:
```diff
@@ -1,28 +1,62 @@
+    setIsHovering(true);
+  };
+
+  const premiumSticker = () => {
+    if (premium === true) {
+      return (
+        <div className="place-card__mark">
+          <span>Premium</span>
+        </div>
+      );
```
уже выше писал, в отдельные компонент выноси

saimon395 commented on project/src/components/card.tsx:
```diff
@@ -1,28 +1,62 @@
+  const {offer} = props;
+  const {price, type, rating, title, previewImage, premium} = offer;
+
+  const [ , setIsHovering] = useState(false);
+
+  const handleMouseOver = () => {
+    setIsHovering(true);
```
вообще по заданию стэйт повыше должен быть, а именно в компоненте списке, то есть в OffersList

saimon395 replied:
плюс положи этот компонент в отдельную папку

saimon395 commented on project/src/components/app/app.tsx:
```diff
@@ -31,7 +34,7 @@ function App({offerCount}: AppWelcomePageProps): JSX.Element {
           <Route
             path={AppRoute.Offer}
-            element = {<OfferUserLoggedPage/>}
+            element = {<OfferUserLoggedPage offers={offers[0]}/>}
```
вот здесь нужно не первый оффер передавать а все, а уже внутри компонента с помощью useParams доставать из адресной строки id и искать по нему оффер в общем списке

saimon395 commented on project/src/components/app/app.tsx:
```diff
@@ -1,25 +1,28 @@
         <Routes>
           <Route
             path={AppRoute.Root}
-            element = {<WelcomePage offerCount={offerCount}/>}
+            element = {<WelcomePage offerCount={offerCount} offers = {offers}/>}
           />
           <Route
             path={AppRoute.Root}
```
плюс убирай повторяющиеся роуты, от них всё равно толку нет, будет только первый работать

saimon395 commented on project/src/components/card.tsx:
```diff
@@ -1,28 +1,62 @@
+  offer: Offer;
+}
+
+function OfferCard(props: OfferUserLoggedPageProps): JSX.Element {
+  const {offer} = props;
+  const {price, type, rating, title, previewImage, premium} = offer;
+
+  const [ , setIsHovering] = useState(false);
+
+  const handleMouseOver = () => {
+    setIsHovering(true);
+  };
+
+  const premiumSticker = () => {
+    if (premium === true) {
+      return (
+        <div className="place-card__mark">
+          <span>Premium</span>
+        </div>
+      );
+    }
+    else {
+      return '';
+    }
+  };
+
+  const star = () => `${Math.floor(rating * 2) * 10}%`;
+
   return (
-    <article className="cities__card place-card">
-      <div className="cities__image-wrapper place-card__image-wrapper">
+    <article className="cities__card place-card"
+      onMouseOver={handleMouseOver}
+    >
+      {premiumSticker()}
+      <div className="cities__image-wrapper place-card__image-wrapper" >
         <a href="#todo">
-          <img className="place-card__image" src="img/room.jpg" width="260" height="200" alt="Place"/>
+          <img className="place-card__image" src={previewImage} width="260" height="200" alt="Place"/>
```
не забудь ссылке адрес на правильный поменять

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -1,7 +1,100 @@
 import Logo from '../../components/logo/logo';
+import { Offers } from '../../types/offer';
+
+type OfferUserLoggedPageProps = {
+  offers: Offers[0];
+}
```
очень странная типизация, у тебя же есть тип Offer, используй его

Ivan-Pelipenko committed:
- Рефакторинг
- Создание компонента "Форма отпраки комментария"

Ivan-Pelipenko committed:
- Сохранение введенных данных в состояние компонента отправки формы

saimon395 commented on project/src/pages/offers-list/offers-list.tsx:
```diff
@@ -0,0 +1,28 @@
+
+  const handleMouseOver = () => {
+    setIsHovering(true);
+  };
```
а зачем ты сэтаешь в стэйт true? там же по заданию должен быть айдишник оффера

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -53,187 +68,68 @@ function OfferUserLoggedPage(): JSX.Element {
-                <span>Premium</span>
-              </div>
+              {premiumSticker(premium)}
               <div className="property__name-wrapper">
```
все функции которые возвращают JSX.Element в реакте - компоненты, а компоненты подключают через треугольные скобки, плюс про именование не забывай, у компонентов названия с большой буквы

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -1,7 +1,22 @@
+  const userId = Number(param.id);
+
+  const offerFromOffers = offers.find((offer) => offer.id === userId);
+  const offer = Object(offerFromOffers);
```
во первых название offerFromOffers не очень хорошее, назови лучше как нибуть типа activeOffer, или currentOffer, а во вторых не плохо бы проверить что тебе find вернул не undefined, потому что в принципе аользователь же может в адресной строке вместо id вписать что угодно, тогда и приведение через Object не понадобится

saimon395 commented on project/src/components/utils/offer-util.tsx:
```diff
@@ -0,0 +1,88 @@
+
+export const stars = (item: number) => `${Math.round(item) * 20}%`;
+
+export const premiumSticker = (item: boolean) => {
+  if (item === true) {
+    return (
+      <div className="property__mark">
+        <span>Premium</span>
+      </div>
+    );
+  }
+  else {
```
всё что JSX.Element возвращает в отдельнве файлы выноси и складывай в папке components , ну и про именование уже писал

saimon395 commented on project/src/components/comment-submission-form/comment-submission-form.tsx:
```diff
@@ -0,0 +1,68 @@
+    review: '',
+  });
+
+  const fieldChangeHandle = (evt: { target: { name: string; value: string } }) => {
+    const {name, value} = evt.target;
+    setFormData({...formData, [name]: value});
+  };
```
не стоит так делать, без крайней нужды, все браузерные события уже типизированы , если не знаешь какой тип использовать гугли )
```suggestion
  const fieldChangeHandle = (evt: React.ChangeEvent<HTMLInputElement>) => {
    const {name, value} = evt.target;
    setFormData({...formData, [name]: value});
  };
```

saimon395 replied:
с остальными обработчиками аналогично

saimon395 commented on project/src/components/card/card.tsx:
```diff
@@ -0,0 +1,43 @@
+
+  return (
+    <>
+      {premiumSticker(premium)}
+      <div className="cities__image-wrapper place-card__image-wrapper" >
```
про компоненты уже писал

Ivan-Pelipenko committed:
Рефакторинг
- Испавлен переход на нужную страницу
- Исправлено состояние "активной карточки"
- Каждый компонент в своей директории
- Исправления по мелочам

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -127,7 +153,35 @@ function OfferUserLoggedPage({offers}: OfferUserLoggedPageProps): JSX.Element {
-                  {reviewsList(reviews)}
+
+                  {reviews.map((review) => {
+                    const keyValue = `${review.avatar}`;
+                    return (
+                      <li key={keyValue}className="reviews__item">
```
здесь можно было отдельную переменную и не создавать, а сразу в key писать review.avatar,

saimon395 replied:
но вообще это плохой вариант для ключа, потому что аватарки могут и повторятся, лучше айдишник используй

saimon395 commented on project/src/pages/offer-user-logged-page/offer-user-logged-page.tsx:
```diff
@@ -68,20 +76,29 @@ function OfferUserLoggedPage({offers}: OfferUserLoggedPageProps): JSX.Element {
-              {collage(images)}
+
+              {images.map((image) => {
+                const keyValue = `${image}`;
+                return (
+                  <div key={keyValue} className="property__image-wrapper">
```
здесь и ниже аналогично, в отдельных переменных смысла нет

Ivan-Pelipenko committed:
- Изменение 'key' в елементах

Ivan-Pelipenko committed:
Задание 5 часть 1
- Установка пакета 'leaflet' и пакета '@types/leaflet'

Ivan-Pelipenko committed:
Добавление координат в моки
